common/thorhelper/thorread.cpp

```diff
@@ -25,13 +25,152 @@
+class DiskReadMapping : public CInterfaceOf<IDiskReadMapping>
+{
+public:
+    DiskReadMapping(RecordTranslationMode _mode, const char * _format, unsigned _actualCrc, IOutputMetaData & _actual, unsigned _expectedCrc, IOutputMetaData & _expected, unsigned _projectedCrc, IOutputMetaData & _output, const IPropertyTree * _options)
+    : mode(_mode), format(_format), actualCrc(_actualCrc), actualMeta(&_actual), expectedCrc(_expectedCrc), expectedMeta(&_expected), projectedCrc(_projectedCrc), projectedMeta(&_output), options(_options)
+    {}
+
+    virtual const char * queryFormat() const override { return format; }
+    virtual unsigned getActualCrc() const override { return actualCrc; }
+    virtual unsigned getExpectedCrc() const override { return expectedCrc; }
+    virtual unsigned getProjectedCrc() const override { return projectedCrc; }
+    virtual IOutputMetaData * queryActualMeta() const override { return actualMeta; }
+    virtual IOutputMetaData * queryExpectedMeta() const override{ return expectedMeta; }
+    virtual IOutputMetaData * queryProjectedMeta() const override{ return projectedMeta; }
+    virtual const IPropertyTree * queryOptions() const override { return options; }
+    virtual RecordTranslationMode queryTranslationMode() const override { return mode; }
+
+    virtual const IDynamicTransform * queryTranslator() const override
+    {
+        ensureTranslators();
+        return translator;
+    }
+    virtual const IKeyTranslator *queryKeyedTranslator() const override
+    {
+        ensureTranslators();
+        return keyedTranslator;
+    }
+
+    virtual bool matches(const IDiskReadMapping * other) const
+    {
+        return mode == other->queryTranslationMode() && streq(format, other->queryFormat()) &&
+                ((actualCrc && actualCrc == other->getActualCrc()) || (actualMeta == other->queryActualMeta())) &&
+                ((expectedCrc && expectedCrc == other->getExpectedCrc()) || (expectedMeta == other->queryExpectedMeta())) &&
+                ((projectedCrc && projectedCrc == other->getProjectedCrc()) || (projectedMeta == other->queryProjectedMeta())) &&
+                areMatchingPTrees(options, other->queryOptions());
+    }
+
+    virtual bool expectedMatchesProjected() const
+    {
+        return (expectedCrc && expectedCrc == projectedCrc) || (expectedMeta == projectedMeta);
+    }
+
+protected:
+    void ensureTranslators() const;
+
+protected:
+    RecordTranslationMode mode;
+    mutable bool checkedTranslators = false;
+    StringAttr format;
+    unsigned actualCrc;
+    unsigned expectedCrc;
+    unsigned projectedCrc;
+    Linked<IOutputMetaData> actualMeta;
+    Linked<IOutputMetaData> expectedMeta;
+    Linked<IOutputMetaData> projectedMeta;
+    Linked<const IPropertyTree> options;
+    mutable Owned<const IDynamicTransform> translator;
+    mutable Owned<const IKeyTranslator> keyedTranslator;
+};
+
+static CriticalSection translatorCS;
```

Comment:
Bit worried about a single shared critsec like this - especially when every call to queryTranslator will be blocked if someone is creating a translator (even if checkedTranslators is true). Is there some way that could be improved?